src/VisualStudio/Core/Def/Implementation/Experimentation/KeybindingResetDetector.cs

```diff
@@ -0,0 +1,358 @@
+
+                    changedOptions = _workspace.Options.WithChangedOption(KeybindingResetOptions.ReSharperStatus, ReSharperStatus.NotInstalledOrDisabled);
+
+                    _workspace.Options = changedOptions;
+                }
+
+                if (needsReset)
+                {
+                    ShowGoldBar();
+                }
+            }
+            else
+            {
+                UpdateReSharperEnableStatus();
+
+                // We need to monitor for suspend/resume commands, so create and install the command target.
+                var priorityCommandTargetRegistrar = _serviceProvider.GetService(typeof(SVsRegisterPriorityCommandTarget)) as IVsRegisterPriorityCommandTarget;
+                hr = priorityCommandTargetRegistrar.RegisterPriorityCommandTarget(
+                    dwReserved: 0 /* from docs must be 0 */,
+                    pCmdTrgt: this,
+                    pdwCookie: out _priorityCommandTargetCookie);
+
+                // Initialize the OleComponent to listen for modal changes (which will tell us when Tools->Options is closed)
+                _oleComponent = OleComponent.CreateHostedComponent("Keybinding Reset Detector");
+                _oleComponent.ModalStateChanged += OnModalStateChanged;
+
+                if (ErrorHandler.Failed(hr))
+                {
+                    FatalError.ReportWithoutCrash(Marshal.GetExceptionForHR(hr));
+                }
+            }
+        }
+
+        private void UpdateReSharperEnableStatus()
+        {
+            AssertIsForeground();
+
+            var isEnabled = IsReSharperEnabled();
+            var options = _workspace.Options;
+
+            if (isEnabled)
+            {
+                // Update the option if we weren't already enabled. There's no other actions to take, since
+                // ReSharper is enabled and we don't want to pop a gold bar
+                if (options.GetOption(KeybindingResetOptions.ReSharperStatus) != ReSharperStatus.Enabled)
+                {
+                    options = options.WithChangedOption(KeybindingResetOptions.ReSharperStatus, ReSharperStatus.Enabled)
+                                     .WithChangedOption(KeybindingResetOptions.NeedsReset, false);
+                    _workspace.Options = options;
+                }
+            }
+            else
+            {
+                bool needsReset;
+                if (options.GetOption(KeybindingResetOptions.ReSharperStatus) != ReSharperStatus.Suspended)
+                {
+                    options = options.WithChangedOption(KeybindingResetOptions.ReSharperStatus, ReSharperStatus.Suspended)
+                                     .WithChangedOption(KeybindingResetOptions.NeedsReset, true);
+                    needsReset = true;
+                    _workspace.Options = options;
+                }
+                else
+                {
+                    needsReset = options.GetOption(KeybindingResetOptions.NeedsReset);
```

Comment:
when can this happen?

Reply:
If ReSharper was already suspended on the last session, we want to use whatever the existing NeedsReset was.

---
In reply to: [155666991](https://github.com/dotnet/roslyn/pull/23605#discussion_r155666991) [](ancestors = 155666991)